apps/prairielearn/assets/scripts/externalImageCaptureClient.ts

```diff
@@ -0,0 +1,209 @@
+    }
+  }
+
+  function displayImagePreview(dataUrl: string) {
+    if (!imagePreview || !uploadButton || !cameraInputLabelSpan) {
+      throw new Error(
+        'Required elements for displaying an image preview not found in the document',
+      );
+    }
+    imagePreview.src = dataUrl;
+    imagePreview.style.display = 'block';
+
+    uploadButton.disabled = false;
+    cameraInputLabelSpan.textContent = 'Retake photo';
+  }
+
+  function listenForImageCaptureAcknowledgement() {
+    if (!externalImageCaptureForm) {
+      throw new Error('External image capture form not found in the document');
+    }
+    const variant_id = externalImageCaptureForm.dataset.variantId;
+    const file_name = externalImageCaptureForm.dataset.fileName;
+
+    const socket = io('/external-image-capture');
+
+    socket.emit(
+      'joinExternalImageCapture',
+      {
+        variant_id,
+        file_name,
+      },
+      (msg: StatusMessageWithFileContent) => {
+        if (!msg) {
+          changeState('failed');
+          throw new Error('Failed to join external image capture room');
+        }
+      },
+    );
+
+    const timeoutMs = 10 * 1000; // 10 seconds
+
+    const timeout = setTimeout(() => {
+      changeState('failed');
+      socket.disconnect();
+    }, timeoutMs);
+
+    socket.on('externalImageCaptureAck', (msg: StatusMessage) => {
+      clearTimeout(timeout);
+      socket.disconnect();
+
+      if (!msg) {
+        changeState('failed');
+        throw new Error('Failed to receive external image capture acknowledgement');
+      }
+
+      changeState('success');
+    });
+  }
+
+  cameraInput.addEventListener('change', () => {
+    cameraInput.disabled = false;
+    if (cameraInput.files && cameraInput.files.length > 0) {
+      // Select the image the user uploaded.
+      const file = cameraInput.files[0];
+      const url = URL.createObjectURL(file);
+      const image = new Image();
+
+      image.src = url;
+
+      image.onload = () => {
+        // Perform scaling to ensure that user-uploaded images are not too large.
+        // The scale factor ensures that the width and height of the image do not exceed 1000px.
+        // If the image width and height are both less than 1000px, no scaling is applied.
+        const imageScaleFactor = 1000 / Math.max(image.width, image.height);
+
+        if (imageScaleFactor >= 1) {
+          // No scaling is necessary, so we can directly use the original image.
+          displayImagePreview(url);
+          return;
+        }
+
+        const targetWidth = Math.round(image.width * imageScaleFactor);
+        const targetHeight = Math.round(image.height * imageScaleFactor);
+
+        const canvas = document.createElement('canvas');
+        canvas.width = targetWidth;
+        canvas.height = targetHeight;
+
+        const ctx = canvas.getContext('2d');
+        if (!ctx) {
+          throw new Error('Failed to get canvas context');
+        }
```

Comment:
This is a bunch of processing that could be memoized instead of being repeated with every frame.